### ISSUE TITLE
TypeVarTuple and *args Interaction

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#2993]
I'm a little confused on the expected interaction of TypeVarTuple and *args. So unconfident if this is a bug or as expected behavior. Is TypeVarTuple supposed to work with variadic positional argument? My toy example is,

```python
from __future__ import annotations
from typing import Callable, TypeVar
from typing_extensions import TypeVarTuple

Ts = TypeVarTuple('Ts')
R = TypeVar('R')

def map_structure(func: Callable[[*Ts], R], *structure: *Ts) -> R:
    return func(structure)

def my_sum(*args: int) -> int:
    return sum(args)

map_structure(my_sum, 1, 2, 3)

```

This produces a type error of,

> Argument of type "(*args: int) -> int" cannot be assigned to parameter "func" of type "(*Ts@map_structure) -> R@map_structure" in function "map_structure"
>   Type "(*args: int) -> int" cannot be assigned to type "(*Ts@map_structure) -> R@map_structure"P
> 

Not sure what decision on boundedness of typevartuples end up being. The pep description here,

> The behavior of a Callable containing an unpacked item, whether the item is a TypeVarTuple or a tuple type, is to treat the elements as if they were the type for *args. So, Callable[[*Ts], None] is treated as the type of the function:
> 
> def foo(*args: *Ts) -> None: ...
> 

sounds like this should be allowed, but a little unsure. I was expecting 

```python
def my_sum(*args: int) -> int:
    return sum(args)
```

to be compatible with `Callable[*Ts, R]`.

Version of pyright is 1.1.217

**Comment by erictraut** [comment]
PEP 646 is *very* complicated. I predict that it will be a long time before we work out all of the issues and edge cases. And it's going to take an even longer time to get consistent behaviors across all of the major type checkers.

I don't know if the PEP contemplates the use case in your sample. I didn't include support for this case it in my implementation. When I first implemented support for PEP 646, TypeVarTuple didn't support arbitrary-length tuples, but that support was added in the latest version of the PEP. Therefore, I think it should theoretically work to match `*args: int` to `*Ts`.

I'd be interested in input from the authors of the PEP. @mrahtz, @shannonzhu, @pradeep90, is this case intended to work? Does it work in your reference implementation in pyre?

I just tried the latest published version of pyre (0.9.10) to see what it does with this code sample, but it gave me errors when accessing TypeVarTuple. Would you expect it to work?

**Comment by JelleZijlstra** [comment]
I believe pyre would need you to use `from pyre_extensions import TypeVarTuple`. (I haven't tried myself though.)

**Comment by erictraut** [comment]
Yes, importing from `pyre_extensions` worked. Thanks for the suggestion, Jelle.

Pyre (which is the reference implementation for PEP 646) emits an error for the statement `map_structure(my_sum, 1, 2, 3)`:
```
test.py:14:14 Incompatible parameter type [6]: In call `map_structure`, for 1st positional only parameter expected `typing.Callable[[Variable(*test.Ts)], Variable[R]]` but got `typing.Callable(my_sum)[[Variable(int)], int]`.
```

So this is apparently not allowed by the PEP, and pyright is doing the right thing here.

I'm going to close this issue for now. I'll re-open if we get a clarifying response from the PEP authors and they indicate that this is meant to work.

**Comment by mrahtz** [comment]
Thanks for reporting, Eric and hmc-cs-mdrissi!

This is a very reasonable example and should definitely work. The following code (which is slightly tweaked from the original example) works fine in Pyre:

```python
from typing import Callable, cast, TypeVar
from pyre_extensions import TypeVarTuple, Unpack

Ts = TypeVarTuple('Ts')
R = TypeVar('R')

def map_structure(func: Callable[[Unpack[Ts]], R], *structure: Unpack[Ts]) -> R:
    return func(*structure)

def my_sum(*args: Unpack[tuple[int, ...]]) -> int:
    return sum(args)

map_structure(my_sum, 1, 2, 3)
```

The changes I had to do to make this work were:
* In `my_sum`, changing the type from `*args: int` to `*args: *tuple[int, ...]`. These should be exactly the same, so I think this is a Pyre bug; I've reported at https://github.com/facebook/pyre-check/issues/578.
* In `map_structure`, changing `func(structure)` to `func(*structure)`. Technically I think it should work even with the former (we're saying `func` has a single variadic positional argument, and we're passing a single tuple, so should be fine) but a) I'm guessing the latter was what you _meant_ to do, and b) there also seems to be a bug with the former in Pyre, which I've reported at https://github.com/facebook/pyre-check/issues/579.

But yeah, the original example should also work fine - so I think this might be, as Eric says, one of the things we added support for in the PEP after the initial implementation in pyright.

**Comment by erictraut** [comment]
Thanks for the clarification, @mrahtz. Reopening.

**Comment by pradeep90** [comment]
> In map_structure, changing func(structure) to func(*structure). Technically I think it should work even with the former (we're saying func has a single variadic positional argument, and we're passing a single tuple, so should be fine) 

This is not true. `Callable[[*Ts], R]` can accept `func(*structure)` but not `func(structure)`. As explained in the [issue](https://github.com/facebook/pyre-check/issues/579), the latter is unsound. To see why, we can try it out at runtime:

```
>>> def map_structure(func, *structure):
...   return func(structure)
... 
>>> def my_sum(*args):
...   return sum(args)
... 
>>> map_structure(my_sum, 1, 2, 3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in map_structure
  File "<stdin>", line 2, in my_sum
TypeError: unsupported operand type(s) for +: 'int' and 'tuple'


# Works with `*structure`
>>> def map_structure(func, *structure):
...   return func(*structure)
... 
>>> map_structure(my_sum, 1, 2, 3)
6
```

Overall, the code snippet Matthew shared should work without trouble, as per PEP 646. (The `*args: *Tuple[int, ...]` vs `*args: int` discrepancy is indeed a Pyre bug.)

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.220, which I just published. It will also be included in the next release of pylance.
